src/Catel.Core/Catel.Core.Shared/Services/ObjectIdGenerator.cs

```diff
@@ -51,12 +55,52 @@ public void ReleaseIdentifier(TUniqueIdentifier identifier)
             }
         }
 
+        /// <inheritdoc />
+        public TUniqueIdentifier GetUniqueIdentifierForInstance(TObjectType instance, bool reuse = false)
+        {
+            Argument.IsNotNull("instance", instance);
+
+            lock (_syncObj)
+            {
+                if (_allocatedUniqueIdentifierPerInstances.TryGetValue(instance, out var wrapper))
+                {
+                    return wrapper.UniqueIdentifier;
+                }
+
+                wrapper = new InstanceWrapper(this, GetUniqueIdentifier(reuse));
+
+                _allocatedUniqueIdentifierPerInstances.Add(instance, wrapper);
+
+                return wrapper.UniqueIdentifier;
+            }
+        }
+
         /// <summary>
         /// Generates the unique identifier.
         /// </summary>
         /// <returns>
         /// The unique identifier.
         /// </returns>
         protected abstract TUniqueIdentifier GenerateUniqueIdentifier();
+
+        private class InstanceWrapper
+        {
+            public InstanceWrapper(IObjectIdGenerator<TObjectType, TUniqueIdentifier> objectIdGenerator, TUniqueIdentifier uniqueIdentifier)
+            {
+                Argument.IsNotNull("objectIdGenerator", objectIdGenerator);
```

Comment:
We can remove this since this is a private class. Then it saves us a bit performance as well.